syntax = "proto3";

package arduino_trader.gateway;

import "common/common.proto";

// Gateway Service - Orchestrates workflows and provides API gateway
service GatewayService {
  // Get system status
  rpc GetSystemStatus(GetSystemStatusRequest) returns (GetSystemStatusResponse);

  // Trigger full trading cycle
  rpc TriggerTradingCycle(TriggerTradingCycleRequest) returns (stream TradingCycleUpdate);

  // Trigger deposit processing
  rpc ProcessDeposit(ProcessDepositRequest) returns (ProcessDepositResponse);

  // Get job status
  rpc GetJobStatus(GetJobStatusRequest) returns (GetJobStatusResponse);

  // List all jobs
  rpc ListJobs(ListJobsRequest) returns (ListJobsResponse);

  // Trigger specific job
  rpc TriggerJob(TriggerJobRequest) returns (TriggerJobResponse);

  // Get service health
  rpc GetServicesHealth(GetServicesHealthRequest) returns (GetServicesHealthResponse);

  // Health check
  rpc HealthCheck(common.Empty) returns (common.HealthCheckResponse);
}

// Get system status request
message GetSystemStatusRequest {
  bool include_details = 1;
}

// Get system status response
message GetSystemStatusResponse {
  string status = 1;  // "healthy", "degraded", "down"
  SystemMetrics metrics = 2;
  repeated ServiceHealth services = 3;
  repeated JobStatus jobs = 4;
  common.Timestamp last_updated = 5;
}

// System metrics
message SystemMetrics {
  int64 uptime_seconds = 1;
  double cpu_usage_pct = 2;
  double memory_usage_pct = 3;
  int64 total_trades = 4;
  int64 active_positions = 5;
  common.Money total_portfolio_value = 6;
}

// Service health
message ServiceHealth {
  string service_name = 1;
  bool healthy = 2;
  string status = 3;
  string address = 4;
  int32 port = 5;
  double latency_ms = 6;
  common.Timestamp last_check = 7;
}

// Job status
message JobStatus {
  string job_name = 1;
  string status = 2;  // "idle", "running", "completed", "failed"
  common.Timestamp last_run = 3;
  common.Timestamp next_run = 4;
  int32 run_count = 5;
  int32 success_count = 6;
  int32 failure_count = 7;
  string last_error = 8;
}

// Trigger trading cycle request
message TriggerTradingCycleRequest {
  bool dry_run = 1;
  bool force = 2;
}

// Trading cycle update (streaming)
message TradingCycleUpdate {
  string step = 1;  // "syncing", "planning", "executing", "complete"
  int32 progress_pct = 2;
  string message = 3;
  bool complete = 4;
  TradingCycleResult result = 5;
  string error = 6;
}

// Trading cycle result
message TradingCycleResult {
  bool success = 1;
  int32 trades_executed = 2;
  int32 trades_failed = 3;
  common.Money total_invested = 4;
  common.Money total_cost = 5;
  repeated TradeResult trades = 6;
}

// Trade result
message TradeResult {
  string trade_id = 1;
  string symbol = 2;
  common.TradeSide side = 3;
  double quantity = 4;
  common.Money price = 5;
  bool success = 6;
  string error = 7;
}

// Process deposit request
message ProcessDepositRequest {
  common.Money amount = 1;
  bool auto_invest = 2;
}

// Process deposit response
message ProcessDepositResponse {
  bool success = 1;
  string transaction_id = 2;
  common.Money new_cash_balance = 3;
  string message = 4;
}

// Get job status request
message GetJobStatusRequest {
  string job_name = 1;
}

// Get job status response
message GetJobStatusResponse {
  bool found = 1;
  JobStatus job = 2;
}

// List jobs request
message ListJobsRequest {
  string status_filter = 1;  // Optional filter
}

// List jobs response
message ListJobsResponse {
  repeated JobStatus jobs = 1;
}

// Trigger job request
message TriggerJobRequest {
  string job_name = 1;
  bool force = 2;
  map<string, string> parameters = 3;
}

// Trigger job response
message TriggerJobResponse {
  bool success = 1;
  string message = 2;
  string run_id = 3;
}

// Get services health request
message GetServicesHealthRequest {
  bool force_check = 1;
}

// Get services health response
message GetServicesHealthResponse {
  bool all_healthy = 1;
  repeated ServiceHealth services = 2;
  int32 healthy_count = 3;
  int32 unhealthy_count = 4;
}
